packages/compose/src/hooks/use-resize-observer/index.ts

```diff
@@ -0,0 +1,186 @@
+ * useResizeObserver(
+ * 	// ...
+ * 	{
+ * 		targetElement,
+ * 		// ...
+ * 	}
+ * );
+ * <div ref={ setTargetElement } />;
+ *
+ * // The element could be obtained through other means, for example:
+ * useEffect( () => {
+ * 	const element = document.querySelector(
+ * 		`[data-element-id="${ elementId }"]`
+ * 	);
+ * 	setTargetElement( element );
+ * }, [ elementId ] );
+ * ```
+ *
+ * @param onUpdate              Callback that will be called when the element is measured (initially and after resizes).
+ * @param options               Options that, with the exception of `targetElement`, will be passed to `ResizeObserver.observe` when called internally. Updating them will not cause the observer to be re-created, and they will only take effect if a new element is observed.
+ * @param options.targetElement The target element to observe. This parameter is an alternative to the returned ref. The element can be changed dynamically.
+ */
+export default function useResizeObserver< T extends HTMLElement >(
+	/**
+	 * Callback that will be called when the element is measured (initially and
+	 * after resizes).
+	 */
+	onUpdate: (
+		/**
+		 * The list of
+		 * [`ResizeObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry)
+		 * objects passed to the `ResizeObserver.observe` callback internally.
+		 */
+		resizeObserverEntries: ResizeObserverEntry[],
+		/**
+		 * The element being tracked at the time of this update.
+		 */
+		element: T
+	) => void,
+	/**
+	 * Options that, with the exception of `targetElement`, will be passed to
+	 * `ResizeObserver.observe` when called internally.
+	 *
+	 * Updating them will not cause the observer to be re-created, and they will
+	 * only take effect if a new element is observed.
+	 */
+	options?: ObserveElementSizeOptions< T >
+): ( element?: T | null | undefined ) => void;
+
+/**
+ * **This is a legacy API and should not be used.**
+ *
+ * @deprecated Use the other `useResizeObserver` API instead: `const ref = useResizeObserver( ( entries ) => { ... } )`.
+ *
+ * Hook which allows to listen to the resize event of any target element when it changes size.
+ * _Note: `useResizeObserver` will report `null` sizes until after first render.
+ *
+ * @example
+ *
+ * ```js
+ * const App = () => {
+ * 	const [ resizeListener, sizes ] = useResizeObserver();
+ *
+ * 	return (
+ * 		<div>
+ * 			{ resizeListener }
+ * 			Your content here
+ * 		</div>
+ * 	);
+ * };
+ * ```
+ */
+export default function useResizeObserver(): [ ReactElement, ObservedSize ];
+
+export default function useResizeObserver< T extends HTMLElement >(
+	onUpdate?: (
+		resizeObserverEntries: ResizeObserverEntry[],
+		element: T
+	) => void,
+	options: ObserveElementSizeOptions< T > = {}
+):
+	| ( ( element?: T | null | undefined ) => void )
+	| [ ReactElement, ObservedSize ] {
+	return onUpdate
+		? _useResizeObserver( onUpdate, options )
+		: _useLegacyResizeObserver();
```

Comment:
In a strict sense this is an illegal hook because it's not able to handle an update from `useResizeObserver( onUpdate )` to `useResizeObserver()`. It will probably crash. In `useSelect` we have a similar situation and handle it by detecting such an unsupported update and throwing an error:

https://github.com/WordPress/gutenberg/blob/trunk/packages/data/src/components/use-select/index.js#L299-L310

We should also:
1. Issue a `deprecated` warning when `useResizeObserver` is used in a legacy way.
2. Export the old version under a different name that doesn't say "legacy". It's still a useful and widely used utility, the only thing deprecated about it is the name, nothing else.

Reply:
We can't issue a `deprecated` warning unless we migrate all of our core usage.

Reply:
I was thinking whether to detect a dynamic change from one way of using it to another, but honestly that'd already be a type error and it'd be a very convoluted thing to write (conditionally passing a callback?).

So I say let it crash if the user does so. No need to safeguard against blatant misuse and we save some extra bloat that's unlikely to ever run.

Reply:
Regarding the deprecated stuff, I feel like at least for now, the jsdocs for the legacy overload are decent enough.

Regarding re-exporting I would say that I find the legacy strategy very unergonomic since it depends on passing a child component that adds constraints around styling. I would propose instead creating an abstraction that has the same API, except that instead of (or in addition to) taking a callback it returns {ref, size}.

That way we can still reliably measure elements with a nicer API and pretty similar usage (main change being that instead of passing a child you pass a ref).

That'd also be quite simple to migrate in core, and we can then deprecate while also announcing the other utility. What do you guys think?

Reply:
> Regarding re-exporting I would say that I find the legacy strategy very unergonomic since it depends on passing a child component that adds constraints around styling. I would propose instead creating an abstraction that has the same API, except that instead of (or in addition to) taking a callback it returns {ref, size}.

To be honest, I'm not even sure that we need that abstraction. Can we try to see how much we can migrate to the new API without intermediary APIs/abstractions. 

I agree though that for now, soft-deprecating (comment in the code) would be a good step. I only see 6 existing usage of the hook in our code base, hopefully, the migration wouldn't be that hard until we can fully deprecate.

Reply:
I see another 11 usages in Calypso. But it shouldn't be hard to migrate. You replace:
```js
const [ observer, size ] = useResizeObserver();
return (
  <div>
    { observer }
    { ... }
  </div>
);
```
with
```js
const ref = useResizeObserver( onResize );
return (
  <div ref={ ref }>
    { ... }
  </div>
);
```
There is always some parent `div` to which you can attach the `ref`, and you always wanted to know the size of _that_ `div` anyway.

But to do the replacement fully, you need to also implement the `onResize` callback that extracts the size and stores it into state. And at least the `extractSize` utility should be shared. It has all the logic to deal with legacy and buggy browsers. On the other hand, it's likely that all browsers today have a correct implementation, and `entry.contentBoxSize[0]` works everywhere. For example, there used to be a bug in Firefox where the `contentBoxSize` was not an array, but that's been [fixed](https://bugzilla.mozilla.org/show_bug.cgi?id=1599176) for 3 years.

Reply:
Worth investigating in a follow-up.